### PR TITLE
Add Parquet loader + from_parquet and to_parquet

### PR DESCRIPTION
**Pull request by lhoestq**
Continuation of #2247 

I added a "parquet" dataset builder, as well as the methods `Dataset.from_parquet` and `Dataset.to_parquet`.
As usual, the data are converted to arrow in a batched way to avoid loading everything in memory.

### CODE REVIEW

**Comment by lhoestq** [comment]
`pyarrow` 1.0.0 doesn't support some types in parquet, we'll have to bump its minimum version.

Also I still need to add dummy data to test the parquet builder.

**Comment by lhoestq** [comment]
I had to bump the minimum pyarrow version to 3.0.0 to properly support parquet.

File: tests/test_dataset_dict.py [hunk 1]
```diff
@@ -587,6 +587,62 @@ def test_datasetdict_from_json_splits(split, jsonl_path, tmp_path):
...
+        None,
+        {"col_1": "string", "col_2": "int64", "col_3": "float64"},
+        {"col_1": "string", "col_2": "string", "col_3": "string"},
+        {"col_1": "int32", "col_2": "int32", "col_3": "int32"},
+        {"col_1": "float32", "col_2": "float32", "col_3": "float32"},
+    ],
+)
+def test_datasetdict_from_parquet_features(features, parquet_path, tmp_path):
+    cache_dir = tmp_path / "cache"
+    # CSV file loses col_1 string dtype information: default now is "int64" instead of "string"
```
Comment (albertvillanova) [inline tests/test_dataset_dict.py:622]: And the same here.

File: src/datasets/arrow_dataset.py [hunk 2]
```diff
@@ -519,6 +519,45 @@ def from_json(
...
+        columns: Optional[List[str]] = None,
+        **kwargs,
+    ):
+        """Create Dataset from Parquet file(s).
+
+        Args:
+            path_or_paths (path-like or list of path-like): Path(s) of the Parquet file(s).
+            split (:class:`NamedSplit`, optional): Split name to be assigned to the dataset.
+            features (:class:`Features`, optional): Dataset features.
+            cache_dir (:obj:`str`, optional, default ``"~/datasets"``): Directory to cache data.
```
Comment (albertvillanova) [inline src/datasets/arrow_dataset.py:538]: Same wrong default cache_dir.

File: src/datasets/dataset_dict.py [hunk 3]
```diff
@@ -789,6 +789,37 @@ def from_json(
...
+
+        Args:
+            path_or_paths (dict of path-like): Path(s) of the CSV file(s).
+            features (:class:`Features`, optional): Dataset features.
+            cache_dir (str, optional, default="~/datasets"): Directory to cache data.
+            keep_in_memory (bool, default=False): Whether to copy the data in-memory.
+            columns (:obj:`List[str]`, optional): If not None, only these columns will be read from the file.
+                A column name may be a prefix of a nested field, e.g. 'a' will select
+                'a.b', 'a.c', and 'a.d.e'.
+            **kwargs: Keyword arguments to be passed to :meth:`pandas.read_csv`.
```
Comment (albertvillanova) [inline src/datasets/dataset_dict.py:811]: Here you can remove :meth:`pandas.read_csv`.

**Review by albertvillanova — APPROVED** [review]
Thanks @lhoestq.

Only some small comments/questions...

File: setup.py [hunk 4]
```diff
@@ -73,9 +73,9 @@
     # We use numpy>=1.17 to have np.random.Generator (Dataset shuffling)
     "numpy>=1.17",
     # Backend and serialization.
-    # Minimum 1.0.0 to avoid permission errors on windows when using the compute layer on memory mapped data
+    # Minimum 3.0.0 to support mix of struct and list types in parquet format
     # pyarrow 4.0.0 introduced segfault bug, see: https://github.com/huggingface/datasets/pull/2268
-    "pyarrow>=1.0.0,!=4.0.0",
+    "pyarrow>=3.0.0,!=4.0.0",
```
[the remark this answers is not in the document]
Comment (lhoestq) [reply setup.py:78]: Good point, I just checked and there are still many projects that use pyarrow 2.0.0 and 1.0.1
I'll make the change

File: tests/utils.py [hunk 5]
```diff
@@ -34,6 +35,19 @@ def parse_flag_from_env(key, default=False):
...
+def require_pyarrow_at_least_3(test_case):
+    """
+    Decorator marking a test that requires PyArrow 3.0.0
+    to allow nested types in parquet, as well as batch iterators of parquet files.
+
+    These tests are skipped when the PyArrow version is outdated.
+
+    """
+    if version.parse(config.PYARROW_VERSION) < version.parse("3.0.0"):
+        test_case = unittest.skip("test requires PyTorch")(test_case)
```
Comment (albertvillanova) [inline tests/utils.py:47]: You could fix the skip message.

File: .circleci/config.yml [hunk 6]
```diff
@@ -116,8 +116,8 @@ workflows:
         jobs:
             - check_code_quality
             - run_dataset_script_tests_pyarrow_latest
-            - run_dataset_script_tests_pyarrow_1
+            - run_dataset_script_tests_pyarrow_3
```
Comment (albertvillanova) [inline .circleci/config.yml:119]: You should undo this edit too, to fix CI.

File: .circleci/config.yml [hunk 7]
```diff
@@ -116,8 +116,8 @@ workflows:
...
             - run_dataset_script_tests_pyarrow_latest_WIN
-            - run_dataset_script_tests_pyarrow_1_WIN
+            - run_dataset_script_tests_pyarrow_3_WIN
```
Comment (albertvillanova) [inline .circleci/config.yml:121]: And undo this one also.

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Some little fixes.

File: .circleci/config.yml (hunk 6 above)
Comment (lhoestq) [reply .circleci/config.yml:119]: Whoops indeed, thanks

**Comment by lhoestq** [comment]
Done !
Now we're still allowing pyarrow>=1.0.0, but when users want to use parquet features they're asked to update to pyarrow>=3.0.0

**Pull request merged**